### DISCUSSION TITLE
Should we keep database adapters?

### DISCUSSION THREAD

**Discussion by frankie567** [fastapi-users/fastapi-users#913] [General]
Hi there 👋

In this discussion, I would like to have **your** feedback, as a FastAPI Users... user 😄 Your help will be greatly appreciated and will help to shape the future of the library.

## Context

As you know, since the beginning of this library, one of the objective was be 100% database-agnostic to allow developers to plug the DBMS or ORM of their choice. Today, we officially support SQLAlchemy, MongoDB and Tortoise ORM. It's also quite easy for you to develop your own adapter if needed.

For some weeks now, I started to question this choice and wondered if it's necessary to keep this feature.

## Numbers

Looking at the number of downloads from PyPI, we clearly see a vast majority of users actually use SQLAlchemy. Admittedly, this is a partial view of the actual usage, but it already gives a good overview.

![Share of downloads for each DB adapters](https://user-images.githubusercontent.com/1144727/153140612-7eb4cd9f-72df-4be2-b9c1-0a56f690de1e.svg)

## Pros and cons

Of course, the main advantage of this feature is to allow developers to use FastAPI Users whatever their stack is.

However it comes at a quite heavy price:

* Slow down new feature additions: implementing things like roles, refresh tokens are made complex because I always need to think about the interaction with the different database and add adapters all along the way.
* Installation and configuration is difficult: with different installation paths given your adapter, with some having some weirdnesses (like Tortoise ORM interaction with Pydantic), it looks like people are having a hard time installing the library. Lot of questions on the discussions are just because of a misunderstanding of the installation path. This point was also raised in #880.
* Maintenance burden: more of a personal point, but it's quite time-consuming to maintain each sub-library, docs and examples for each adapter.

## Homework for you

Now that you have all the context, I'll ask for your help. Here is the two things I would like you to do.

### 1. Answer the poll

By adding a reaction to this message, tell me which adapter you're using, with the following convention:

* ❤️ SQLAlchemy
* 🚀 MongoDB
* 👀 Tortoise ORM
* 😕 Other / custom

### 2. Give your insights

I would be glad if you could take the time to give your insights about this, especially if you think it's necessary to keep the database adapter feature.

My decision is not made, so it'll be with your help that we'll take the best one 😄 Cheers!

**Comment by justb4** [comment]
Thanks for this @frankie567! I see the dilemma, me also maintaining multi-backend Open Source projects.

Using SQLAlchemy/Alembic, mainly historic (Flask) reasons. `async` and `Typing` give enough challenges, for the good, to level-up my Python skills 🤓 . Actually also using "Custom" (ticked both): keeping data in a Payment Provider backend. The User Table is the only table with one extra field: a unique key for a "Customer" in the Payment Provider API. So probably could be 100% Custom (as the P. Provider has user-defined attrs)...And using [FastAPI-Mail](https://sabuhish.github.io/fastapi-mail) 👍 , think most will need to use email in some form (?).

**Reply by frankie567** [reply]
Thank you for your reply @justb4! 

To be sure there is no misunderstanding here, the question here is to decide if we should keep support for different DBMS/ORM, like MongoDB or Tortoise.

What you seem to describe here is the possibility to add new properties to the user object, like name or phone number for example. This is something I definitely intend to keep, it's crucial for most applications :)

**Reply by justb4** [reply]
I diverted too much: I was not promoting to extend the user object, just showing how I am applying `FastAPI-Users`. 
So all in all: would vote for SQLAlchemy and hooks for Custom backend.

**Comment by frankie567** [comment]
So, right now I've not enough feedbacks and data to make a good decision. I'll postpone it, but I still invite you to give your insights about this.
